spanner/metrics.go

```diff
@@ -0,0 +1,474 @@
+	monitoredResLabelKeyProject        = "project_id"
+	monitoredResLabelKeyInstance       = "instance_id"
+	monitoredResLabelKeyInstanceConfig = "instance_config"
+	monitoredResLabelKeyLocation       = "location"
+
+	// Metric labels
+	metricLabelKeyClientUID         = "client_uid"
+	metricLabelKeyClientName        = "client_name"
+	metricLabelKeyDatabase          = "database"
+	metricLabelKeyMethod            = "method"
+	metricLabelKeyStatus            = "status"
+	metricLabelKeyDirectPathEnabled = "directpath_enabled"
+	metricLabelKeyDirectPathUsed    = "directpath_used"
+
+	// Metric names
+	metricNameOperationLatencies = "operation_latencies"
+	metricNameAttemptLatencies   = "attempt_latencies"
+	metricNameOperationCount     = "operation_count"
+	metricNameAttemptCount       = "attempt_count"
+
+	// Metric units
+	metricUnitMS    = "ms"
+	metricUnitCount = "1"
+)
+
+// These are effectively const, but for testing purposes they are mutable
+var (
+	// duration between two metric exports
+	defaultSamplePeriod = 5 * time.Minute
+
+	clientName = fmt.Sprintf("spanner-go/%v", internal.Version)
+
+	bucketBounds = []float64{0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 8.0, 10.0, 13.0, 16.0, 20.0, 25.0, 30.0, 40.0,
+		50.0, 65.0, 80.0, 100.0, 130.0, 160.0, 200.0, 250.0, 300.0, 400.0, 500.0, 650.0,
+		800.0, 1000.0, 2000.0, 5000.0, 10000.0, 20000.0, 50000.0, 100000.0, 200000.0,
+		400000.0, 800000.0, 1600000.0, 3200000.0}
+
+	// All the built-in metrics have same attributes except 'status' and 'streaming'
+	// These attributes need to be added to only few of the metrics
+	metricsDetails = map[string]metricInfo{
+		metricNameOperationCount: {
+			additionalAttrs: []string{
+				metricLabelKeyStatus,
+			},
+			recordedPerAttempt: false,
+		},
+		metricNameOperationLatencies: {
+			additionalAttrs: []string{
+				metricLabelKeyStatus,
+			},
+			recordedPerAttempt: false,
+		},
+		metricNameAttemptLatencies: {
+			additionalAttrs: []string{
+				metricLabelKeyStatus,
+			},
+			recordedPerAttempt: true,
+		},
+		metricNameAttemptCount: {
+			additionalAttrs: []string{
+				metricLabelKeyStatus,
+			},
+			recordedPerAttempt: true,
+		},
+	}
+
+	// Generates unique client ID in the format go-<random UUID>@<hostname>
+	generateClientUID = func() (string, error) {
+		hostname := "localhost"
+		hostname, err := os.Hostname()
+		if err != nil {
+			return "", err
+		}
+		return uuid.NewString() + "@" + strconv.FormatInt(int64(os.Getpid()), 10) + "@" + hostname, nil
+	}
+
+	exporterOpts = []option.ClientOption{}
+)
+
+type metricInfo struct {
+	additionalAttrs    []string
+	recordedPerAttempt bool
+}
+
+type builtinMetricsTracerFactory struct {
+	enabled             bool
+	isDirectPathEnabled bool
+
+	// To be called on client close
+	shutdown func()
+
+	// attributes that are specific to a client instance and
+	// do not change across different function calls on client
+	clientAttributes []attribute.KeyValue
+
+	operationLatencies metric.Float64Histogram
```

Comment:
So `operation` is the entire round-trip time from a client interface call which can include multiple `attempts`. `Attempt` is for each RPC call. Am I understanding it correctly? If we get two retryable errors and have two retries, does it count 1 or 2 `attempt`?

Reply:
It will count for 2 attempt.
So retry count can be derived from operationCount-attemptCount so decision was to not export retry.
https://docs.google.com/document/d/199avBkkj8ewCHsUYR2dtnnOVdDMTyWg0O-NtHuzWsiQ/edit?resourcekey=0-UydHkoiIv2akT3LNl5NM_w&tab=t.0#bookmark=id.e0pp687p4vm